pretyping/pretyping.ml

```diff
@@ -625,8 +625,17 @@ let rec pretype k0 resolve_tc (tycon : type_constraint) (env : ExtraEnv.t) evdre
         match tycon with
         | Some ty -> ty
         | None ->
-          new_type_evar env evdref loc in
-        { uj_val = e_new_evar env evdref ~src:(loc,k) ~naming ty; uj_type = ty }
+           new_type_evar env evdref loc in
+      let store =
+        if Flags.is_program_mode () then
+          match k with
+          | Evar_kinds.NamedHole _
+          | Evar_kinds.QuestionMark _
+          | Evar_kinds.ImplicitArg (_, _, false) -> Evarsolve.obligation_store
```

Comment:
It's fragile indeed, the good thing is that it is now the only point where that is relevant for obligations, so it will be the only one to change if we use something else to distinguish them